File: sdk/src/types/game.ts
export enum PlayerActionType {
    SMALL_BLIND = "post-small-blind",
    BIG_BLIND = "post-big-blind",
    FOLD = "fold",
    CHECK = "check",
    BET = "bet",
    CALL = "call",
    RAISE = "raise",
    ALL_IN = "all-in",
    MUCK = "muck",
    SIT_IN = "sit-in",
    SIT_OUT = "sit-out",
    SHOW = "show"
}

export enum NonPlayerActionType {
    DEAL = "deal",
    JOIN = "join",
    LEAVE = "leave",
    NEW_HAND = "new-hand",
}

export const AllPlayerActions = { ...PlayerActionType, ...NonPlayerActionType };

export enum GameType {
    CASH = "cash",
    SIT_AND_GO = "sit-and-go",
    TOURNAMENT = "tournament"
}

export enum GameStatus {
    WAITING_FOR_PLAYERS = "waiting-for-players",
    REGISTRATION = "registration",
    IN_PROGRESS = "in-progress",
    FINISHED = "finished"
}

export enum PlayerStatus {
    ACTIVE = "active",
    BUSTED = "busted",
    FOLDED = "folded",
    ALL_IN = "all-in",
    SEATED = "seated",
    SITTING_OUT = "sitting-out",
    SITTING_IN = "sitting-in",
    SHOWING = "showing"
}

export enum TexasHoldemRound {
    ANTE = "ante",
    PREFLOP = "preflop",
    FLOP = "flop",
    TURN = "turn",
    RIVER = "river",
    SHOWDOWN = "showdown",
    END = "end"
}

export type RakeConfig = {
    rakeFreeThreshold: bigint; // Absolute value below which the hand is rake-free
    rakePercentage: number; // Percentage of the pot that is the rake (e.g., 5 for 5%)
    rakeCap: bigint; // Maximum rake amount (absolute cap)
};

export type RakeConfigDTO = {
    rakeFreeThreshold: string; // Absolute value below which the hand is rake-free
    rakePercentage: number; // Percentage of the pot that is the rake (e.g., 5 for 5%)
    rakeCap: string; // Maximum rake amount (absolute cap)
};

export type GameOptions = {
    minBuyIn: bigint;
    maxBuyIn: bigint;
    minPlayers: number;
    maxPlayers: number;
    smallBlind: bigint;
    bigBlind: bigint;
    timeout: number;
    type: GameType; // Optional for cash games
    rake?: RakeConfig; // Optional rake configuration
    owner?: string; // Table owner who collects rake fees
    otherOptions?: Record<string, any>; // Placeholder for future options
};

export type GameOptionsDTO = {
    minBuyIn?: string;
    maxBuyIn?: string;
    minPlayers?: number;
    maxPlayers?: number;
    smallBlind?: string;
    bigBlind?: string;
    timeout?: number;
    type?: GameType; // Optional for cash games
    rake?: RakeConfigDTO; // Optional rake configuration
    owner?: string; // Table owner who collects rake fees
    otherOptions?: Record<string, any>; // Placeholder for future options
};

// This is the type of the last action of a player
export type ActionDTO = {
    playerId: string,
    seat: number,
    action: PlayerActionType | NonPlayerActionType;
    amount: string;
    round: TexasHoldemRound;
    index: number;
    timestamp: number;
};

export type LegalActionDTO = {
    action: PlayerActionType | NonPlayerActionType;
    min: string | undefined;
    max: string | undefined;
    index: number;
};

export type WinnerDTO = {
    address: string;
    amount: string;
    cards: string[] | undefined;
    name: string | undefined;
    description: string | undefined;
};

export type PlayerDTO = {
    address: string;
    seat: number; // change to position
    stack: string; // BigNumber
    isSmallBlind: boolean,
    isBigBlind: boolean,
    isDealer: boolean,
    holeCards: string[] | undefined;
    status: PlayerStatus;
    lastAction: ActionDTO | undefined;
    legalActions: LegalActionDTO[];
    sumOfBets: string;
    timeout: number;
    signature: string;
};

export type ResultDTO = {
    place: number;
    playerId: string;
    payout: string;
};

export type TexasHoldemGameState = {
    type: string;
    address: string;
    minBuyIn: string;
    maxBuyIn: string;
    minPlayers: number;
    maxPlayers: number;
    smallBlind: string;
    bigBlind: string;
    dealer: number;
    players: string[];
    deck: string;
    communityCards: string[];
    pots: string[];
    lastActedSeat: number;
    actionCount: number;
    handNumber: number;
    round: TexasHoldemRound;
    winners: string[];
    results: ResultDTO[];
    signature: string;
};

export type TexasHoldemStateDTO = {
    type: GameType;
    address: string;
    gameOptions: GameOptionsDTO;
    smallBlindPosition?: number;
    bigBlindPosition?: number;
    dealer?: number;
    players: PlayerDTO[];
    communityCards: string[];
    deck: string; // Assume this will be encrypted
    pots: string[];
    lastActedSeat?: number;
    nextToAct: number;
    previousActions: ActionDTO[];
    actionCount: number;
    handNumber: number;
    round: TexasHoldemRound;
    winners: WinnerDTO[];
    results: ResultDTO[];
    signature: string;
};

export enum SUIT {
    CLUBS = 1,
    DIAMONDS = 2,
    HEARTS = 3,
    SPADES = 4
}

export type Card = {
    suit: SUIT;
    rank: number;
    value: number;
    mnemonic: string;
};

export type GameOptionsResponse = {
    address: string;
    gameOptions: GameOptionsDTO;
};

export type TransactionResponse = {
    nonce: string;
    to: string;
    from: string;
    value: string;
    hash: string;
    signature: string;
    timestamp: string;
    data?: string;
}

export type GameStateResponse = {
    state: TexasHoldemStateDTO;
}

export type PerformActionResponse = {
    state: TexasHoldemStateDTO;
    nonce: string;
    to: string;
    from: string;
    value: string;
    hash: string;
    signature: string;
    timestamp: string;
    data?: string;
}
// Constants for URL parameters
export const KEYS = {
    ACTION_TYPE: "action_type",
    INDEX: "index",
    VALUE: "value",
    SEAT: "seat",
    SEED: "seed"
} as const;

File: pvm/go/types/game.go
package types

import "math/big"

// PlayerActionType represents actions a player can take
type PlayerActionType string

const (
	SmallBlind PlayerActionType = "post-small-blind"
	BigBlind   PlayerActionType = "post-big-blind"
	Fold       PlayerActionType = "fold"
	Check      PlayerActionType = "check"
	Bet        PlayerActionType = "bet"
	Call       PlayerActionType = "call"
	Raise      PlayerActionType = "raise"
	AllIn      PlayerActionType = "all-in"
	Muck       PlayerActionType = "muck"
	Show       PlayerActionType = "show"
)

// NonPlayerActionType represents system actions
type NonPlayerActionType string

const (
	Deal          NonPlayerActionType = "deal"
	DealFlop      NonPlayerActionType = "deal-flop"
	DealTurn      NonPlayerActionType = "deal-turn"
	DealRiver     NonPlayerActionType = "deal-river"
	DealShowdown  NonPlayerActionType = "deal-showdown"
	Winner        NonPlayerActionType = "winner"
	ReturnBet     NonPlayerActionType = "return-bet"
	CollectBlinds NonPlayerActionType = "collect-blinds"
	NewHand       NonPlayerActionType = "new-hand"
	BustOut       NonPlayerActionType = "bust-out"
)

// GameType represents the type of poker game
type GameType string

const (
	CashGame   GameType = "cash-game"
	SitAndGo   GameType = "sit-and-go"
	Tournament GameType = "tournament"
)

// GameStatus represents the current status of the game
type GameStatus string

const (
	WaitingForPlayers GameStatus = "waiting-for-players"
	Registration      GameStatus = "registration"
	InProgress        GameStatus = "in-progress"
	Finished          GameStatus = "finished"
)

// PlayerStatus represents the current status of a player
type PlayerStatus string

const (
	NotActed    PlayerStatus = "not-acted"
	Active      PlayerStatus = "active"
	Folded      PlayerStatus = "folded"
	AllInStatus PlayerStatus = "all-in"
	Busted      PlayerStatus = "busted"
	SittingOut  PlayerStatus = "sitting-out"
	SittingIn   PlayerStatus = "sitting-in"
	Showing     PlayerStatus = "showing"
)

// TexasHoldemRound represents the current betting round
type TexasHoldemRound string

const (
	Ante     TexasHoldemRound = "ante"
	Preflop  TexasHoldemRound = "preflop"
	Flop     TexasHoldemRound = "flop"
	Turn     TexasHoldemRound = "turn"
	River    TexasHoldemRound = "river"
	Showdown TexasHoldemRound = "showdown"
	End      TexasHoldemRound = "end"
)

// GameOptions contains the configuration for a poker game
type GameOptions struct {
	MinBuyIn     *big.Int               `json:"minBuyIn"`
	MaxBuyIn     *big.Int               `json:"maxBuyIn"`
	MinPlayers   int                    `json:"minPlayers"`
	MaxPlayers   int                    `json:"maxPlayers"`
	SmallBlind   *big.Int               `json:"smallBlind"`
	BigBlind     *big.Int               `json:"bigBlind"`
	Timeout      int                    `json:"timeout"`
	Type         GameType               `json:"type"`
	OtherOptions map[string]interface{} `json:"otherOptions,omitempty"`
}

// ActionDTO represents a player action
type ActionDTO struct {
	PlayerID  string           `json:"playerId"`
	Seat      int              `json:"seat"`
	Action    PlayerActionType `json:"action"`
	Amount    string           `json:"amount"`
	Round     TexasHoldemRound `json:"round"`
	Index     int              `json:"index"`
	Timestamp int64            `json:"timestamp"`
}

// LegalActionDTO represents a legal action available to a player
type LegalActionDTO struct {
	Action PlayerActionType `json:"action"`
	Min    string           `json:"min"`
	Max    string           `json:"max"`
	Index  int              `json:"index"`
}

// PlayerDTO represents a player in the game
type PlayerDTO struct {
	Address      string           `json:"address"`
	Seat         int              `json:"seat"`
	Stack        string           `json:"stack"`
	IsSmallBlind bool             `json:"isSmallBlind"`
	IsBigBlind   bool             `json:"isBigBlind"`
	IsDealer     bool             `json:"isDealer"`
	HoleCards    []string         `json:"holeCards"`
	Status       PlayerStatus     `json:"status"`
	LastAction   *ActionDTO       `json:"lastAction,omitempty"`
	LegalActions []LegalActionDTO `json:"legalActions"`
	SumOfBets    string           `json:"sumOfBets"`
	Timeout      int64            `json:"timeout"`
	Signature    string           `json:"signature"`
}

// ResultDTO represents a player's result in the game
type ResultDTO struct {
	Place    int    `json:"place"`
	PlayerID string `json:"playerId"`
	Payout   string `json:"payout"`
}

// WinnerDTO represents a pot winner
type WinnerDTO struct {
	PlayerID string   `json:"playerId"`
	Hand     []string `json:"hand"`
	Rank     string   `json:"rank"`
	PotIndex int      `json:"potIndex"`
	Amount   string   `json:"amount"`
}

// TexasHoldemStateDTO represents the complete game state
type TexasHoldemStateDTO struct {
	Type               GameType         `json:"type"`
	Address            string           `json:"address"`
	GameOptions        GameOptions      `json:"gameOptions"`
	SmallBlindPosition int              `json:"smallBlindPosition"`
	BigBlindPosition   int              `json:"bigBlindPosition"`
	Dealer             int              `json:"dealer"`
	Players            []PlayerDTO      `json:"players"`
	CommunityCards     []string         `json:"communityCards"`
	Deck               string           `json:"deck"`
	Pots               []string         `json:"pots"`
	LastActedSeat      int              `json:"lastActedSeat"`
	ActionCount        int              `json:"actionCount"`
	HandNumber         int              `json:"handNumber"`
	NextToAct          int              `json:"nextToAct"`
	PreviousActions    []ActionDTO      `json:"previousActions"`
	Round              TexasHoldemRound `json:"round"`
	Winners            []WinnerDTO      `json:"winners"`
	Results            []ResultDTO      `json:"results"`
	Signature          string           `json:"signature"`
}
